redirectpolicy: Move the feature behind feature flag and remove CCLRP related code

aditighag opened the pull request:
Check commit description.

Fixes: #13294

joestringer commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -95,13 +95,18 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
+	// CCLRP support will be added in future release.
+	// g.Go(func() error {
+	// 	return createCCLRPCRD(clientset)
+	// })
```
Please remove this code instead.

joestringer replied:
Ditto below for Clusterwide LRP functionality. We should not include commented out code in the codebase.

joestringer commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -95,13 +95,18 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
 	})
 
+	// Reviewers: I'm not sure if it makes sense to wrap this under the feature
+	// flag check since we have related Makefile changes.
```
At a glance, I would assume we should probably not register the CRD if the feature is disabled.

joestringer replied:
Contrary opinion, discuss there: https://github.com/cilium/cilium/pull/13671#discussion_r509806503

joestringer reviewed (changes requested):
Minor nits below.

I briefly considered if there's a way to avoid so many checks against `option.Config.EnableLocalRedirectPolicy` across the codebase. This might be possible via a dummy `k.redirectPolicyManager` implementation but I don't think it makes such a big difference so I'll just mention it in case anyone else is wondering, but I don't think this change is really necessary for this PR.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -581,6 +582,22 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
+		if !found {
+			log.Warn("Without socket lookup kernel functionality, BPF " +
+				"datapath cannot prevent potential loop caused by local-redirect" +
+				"service translation. Needs kernel version >= 5.1")
```
I wonder whether we should move this warning under a `sync.Once()` when adding an LRP. That way, even if you enable LRP, you will only be warned when using the feature. It's possible in a future release we will enable this flag by default, in which case it'd be nice to not automatically warn if the user is running a common kernel like 4.19 (which supports LRP but has this caveat).

christarazi commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -95,13 +95,18 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
 	})
 
+	// Reviewers: I'm not sure if it makes sense to wrap this under the feature
+	// flag check since we have related Makefile changes.
```
I don't think we necessarily need to avoid registering the CRD. There is no harm in registering a CRD that will never have a resource created for it. IMO, we can avoid clutter the codebase with a bunch of ifs and only keep it to a minimum where we truly need it.

christarazi replied:
Also, I didn't quite understand why we are removing it from the Makefile too?

aditighag replied:
Yes, I also feel that we don't necessarily have to avoid registering the CRD (hence the note in the description).  I haven't removed the CLRP related operation from the Makefile (just the CCLRP)?

joestringer replied:
FWIW, I was also initially confused by the Clusterwide LRP changes in this PR since there's no dedicated commit for it and it's not called out in the PR description or commit message (hence unclear whether it's intentional).

christarazi replied:
Cool, so looking over this thread again, I had the same confusion as Joe.

The conclusion here is that we will _not_ register the CLRP CRD conditionally, as there's no harm in it.

christarazi reviewed (changes requested):
Note below on not needing to avoid registering the CLRP CRD. I believe @aanm made a similar change recently where he removed the conditional in which we register the CID CRD.

joestringer commented on Makefile:
```diff
@@ -300,7 +300,6 @@ manifests:
 	mv ${TMPDIR}/cilium.io_ciliumidentities.yaml ./examples/crds/ciliumidentities.yaml
 	mv ${TMPDIR}/cilium.io_ciliumnodes.yaml ./examples/crds/ciliumnodes.yaml
 	mv ${TMPDIR}/cilium.io_ciliumlocalredirectpolicies.yaml ./examples/crds/ciliumlocalredirectpolicies.yaml
-	mv ${TMPDIR}/cilium.io_ciliumclusterwidelocalredirectpolicies.yaml ./examples/crds/ciliumclusterwidelocalredirectpolicies.yaml
```
(Note for other reviewers, this is the Clusterwide LRP which was not implemented)

christarazi commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -95,13 +95,18 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
+	// CCLRP support will be added in future release.
+	// g.Go(func() error {
+	// 	return createCCLRPCRD(clientset)
+	// })
```
From what I have gathered from others on this PR is that CCLRP was never implemented, but we added functions like this anyway? I presume that also explains the Makefile change as well. 

I agree with Joe. Let's remove this CCLRP code and the Makefile change in a dedicated commit. Please also mention in that commit that it will eventually be reverted once the functionality of the CCLRP is implemented.

joestringer replied:
> Please also mention in that commit that it will eventually be reverted once the functionality of the CCLRP is implemented.

Although this would be good detail to include, I'm not sure that this is the case. I had a chat with @aditighag about this a while back and I don't think we were able to point out a specific use case or reason for clusterwide LRP.

christarazi replied:
> > Please also mention in that commit that it will eventually be reverted once the functionality of the CCLRP is implemented.
> 
> Although this would be good detail to include, I'm not sure that this is the case. I had a chat with @aditighag about this a while back and I don't think we were able to point out a specific use case or reason for clusterwide LRP.

Ah OK, I wasn't aware of that. Cool.

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -283,8 +285,10 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
 
-	if lrpNeedsReassign || (oldPodReady != newPodReady) {
-		k.redirectPolicyManager.OnUpdatePod(newK8sPod, lrpNeedsReassign, newPodReady == slim_corev1.ConditionTrue)
+	if option.Config.EnableLocalRedirectPolicy {
```
This should be on line 284

aanm reviewed (changes requested):
Also, I don't see the CLRP being remove from the list of generated CRDs nor being removed from the bindata.

aditighag committed:
redirectpolicy: Remove cluster-wide LRP related code

If needed, it'll be supported in a future release.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
redirectpolicy: Move the feature behind feature flag

Also, warn users about potential service translation caused
loop issue on older kernel versions.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

christarazi commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -98,10 +95,7 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
-	g.Go(func() error {
-		return createCCLRPCRD(clientset)
-	})
+	// }
```
Leftover?

christarazi commented on install/kubernetes/quick-install.yaml:
```diff
@@ -118,6 +118,7 @@ data:
 
   auto-direct-node-routes: "false"
   enable-bandwidth-manager: "false"
+  enable-local-redirect-policy: "false"
```
I think you meant for this to be in the second commit.

christarazi reviewed (changes requested):
Just a few more minor comments to address and we should be good to go for me.

joestringer commented on pkg/redirectpolicy/manager.go:
```diff
@@ -110,6 +114,24 @@ func (rpm *Manager) RegisterGetStores(sg StoreGetter) {
 // internal state with the config fields.
 func (rpm *Manager) AddRedirectPolicy(config LRPConfig) (bool, error) {
+	rpm.warnOnce.Do(func() {
+		if option.Config.EnableLocalRedirectPolicy {
```
nit, is this check necessary? Sure `AddRedirectPolicy()` should not be called if the feature is disabled?

aditighag replied:
Yes, agreed.

joestringer commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -516,8 +516,10 @@ func (k *K8sWatcher) EnableK8sWatcher(ctx context.Context) error {
 
-	// cilium local redirect policies
-	go k.ciliumLocalRedirectPolicyInit(ciliumNPClient)
+	if option.Config.EnableLocalRedirectPolicy {
```
I just noticed, but the other async goroutines for CRD initialization in this function perforrm `asyncControllers.Add(1)` before running a goroutine to initialize the watcher. Presumably this should be done inside the conditional statement here too?

aditighag replied:
Good catch! I didn't mean to run it as a go routine.

joestringer commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -516,8 +516,10 @@ func (k *K8sWatcher) EnableK8sWatcher(ctx context.Context) error {
-	// cilium local redirect policies
-	go k.ciliumLocalRedirectPolicyInit(ciliumNPClient)
+	if option.Config.EnableLocalRedirectPolicy {
+		// cilium local redirect policies
```
This comment is redundant. The line above already says "if ... EnableLocalRedirectPolicy". 
```suggestion
```

aditighag replied:
Well, I'm just following the same style used in the function.

joestringer commented on pkg/k8s/watchers/pod.go:
```diff
@@ -225,7 +225,9 @@ func (k *K8sWatcher) addK8sPodV1(pod *slim_corev1.Pod) error {
 		// GH issue #13136.
-		k.redirectPolicyManager.OnAddPod(pod)
+		if option.Config.EnableLocalRedirectPolicy {
+			k.redirectPolicyManager.OnAddPod(pod)
```
I feel like it'd be a nice followup to get rid of all of the option checks in this file code by replacing the `k.redirectPolicyManager` with a dummy manager if the LRP is not enabled to make absolutely sure there's no paths into the LRP manager if it's disabled. But no need to address in this PR.

aditighag replied:
Yes, I intend to add a dummy manager for unit tests too.

joestringer reviewed (changes requested):
A few minor nits, there's one comment below around asyncControllers which seems like it might be a bug?

aditighag committed:
docs: Update the guide with feature flag instructions

Signed-off-by: Aditi Ghag <aditi@cilium.io>
